app/controllers/dashboards_controller.rb

```diff
@@ -5,16 +5,45 @@ class DashboardsController < ApplicationController
 
   def show
     fetch_and_authorize_user
+
     if params[:which] == "user_followers"
       @follows = Follow.where(followable_id: @user.id, followable_type: "User").
         includes(:follower).order("created_at DESC").limit(80)
     elsif params[:which] == "organization_user_followers"
       @follows = Follow.where(followable_id: @user.organization_id, followable_type: "Organization").
         includes(:follower).order("created_at DESC").limit(80)
     elsif @user&.organization && @user&.org_admin && params[:which] == "organization"
-      @articles = @user.organization.articles.order("created_at DESC").decorate
+      @articles =
+        case params[:sort]
+        when "creation-asc" then @user.organization.articles.order("created_at ASC").decorate
```

Comment:
An idea for refactoring:
```ruby
module Dashboard
  class ArticlesQuery

    ALLOWED_FIELDS = %w[published_at, ...] # list of allowed to sort fields
    # user is actually user or org, probably there's a better name
    def initialize(user, field, order)
      @user = user
      @order = 'desc' unless order == 'asc'
      @field = 'published_at' unless ALLOWED_FIELDS.include?(field)
    end

    def articles
      user.articles.order("#{field} #{order}")
    end
 
    private
    
    attr_reader :user, :order, :field
  end
end
```
Call in the controller like `Dashboard::ArticlesQuery.new(user, field, order).articles`
That requires thoughts on separating `field` and `order`.

Reply:
Really like your refactoring, there's too much logic in the controller right now IMO

> Call in the controller like Dashboard::ArticlesQuery.new(user, field, order).articles. That requires thoughts on separating field and order.

That's why I initially proposed `creation-asc` because it can easily be split in `field` and `order`

Reply:
Thanks. Splitting `field` and `order` seems easy. Then there is just a need to map received field names to a the db field names.

Reply:
oh, and a small fix on a thing that I just noticed: query objects are supposed to take in input the relation, so they can be composable with other external conditions or other query objects, so instead of:

```ruby
Dashboard::ArticlesQuery.new(user, field, order).articles
```

I'd change it to

```ruby
Dashboard::ArticlesQuery.new(user.articles, field, order).sort
```

the query is all about articles, not the user. It also makes them way easier to test because you can pass any "Article" relation instead of the user object 🙂

Reply:
Nice improvement :+1:

Reply:
Here are my various issues with the suggested refactor:

#### Naming
Query in the name implies searching/filtering but be we are doing ordering so so naming it should be sorting or ordering to better describe the functionality.

#### Clever vs Clear
My code is more verbose but it was intentional to be easy to be understood by juniors who I can tell going through the various tickets in this codebase struggle to complete the most basic of tasks which I suspect is due to this codebase deviating from Rails conventions.

#### Convention Over Configuration

There are ways to refactoring this and it would be done as a scope since scoping is already used heavily used in this project

```ruby
scope :sorting, lambda { |value|
  kind, dir = value.split('-')

  return %w{asc desc}.include?(dir)

  column =
  case kind
  when "creation"  then "created_at"
  when "views"     then "page_views_count"
  when "reactions" then "positive_reactions_count"
  when "comments"  then "comments_count"
  when "published" then "published_at"
  end

  order("? ?",column, dir)
}
```

Which would now result in

```ruby
@user.organization.articles.sorting(params[:sort]).decorate

```

I personally don't like the use of any model scoping for maintenance reasons but its the lesser of two abstraction evils.

#### Not Escaping

This here,
`user.articles.order("#{field} #{order}")`
Is not escaping and needs to be as such
`user.articles.order("? ?"),field,order`

#### Test All Conditions Explicity

This only checks in the case of one state ie. `asc and defaulting else` and needs to check both states explicitly.

```ruby
@order = 'desc' unless order == 'asc'
```


```ruby
dir =
case params[:order]
  when 'asc' then "ASC"
  when 'desc' then "DESC"
  else
    'asc'
end
```

another way could have been which is a bit less explicit than the previous example

```ruby
order ||= 'asc'
order = params[:order] if %w[asc desc].include?(param[:order])
```

My appologies for the harsh criticism.

Reply:
An option to explore too would be using [ransack](https://github.com/activerecord-hackery/ransack). I've used it on several projects and it works pretty well.

Reply:
I'm looked at ransack.

My thoughts are it adds another DSL for juniors to come in and worry about.
I think we need a strong use-case beside this one ticket for the addition of such a gem.
This codebase could really use a pruning of gems.

Reply:
> #### Naming
> 
> Query in the name implies searching/filtering but be we are doing ordering so so naming it should be sorting or ordering to better describe the functionality.

I think @lightalloy's was more a suggestion about how to refactor the code, than about the name of the query object itself. 

> #### Clever vs Clear
> 
> My code is more verbose but it was intentional to be easy to be understood by juniors who I can tell going through the various tickets in this codebase struggle to complete the most basic of tasks which I suspect is due to this codebase deviating from Rails conventions.

I agree that code should be kept as simple as possible but that doesn't mean foregoing abstractions and good practices. Having the whole sorting logic inside the controller doesn't make it more readable, just makes the controller longer. You can just can keep your code and put it in a query object that's testable outside the controller for example, if you'd prefer not to be "clever" about using an array or a hash for dispatching.

> I personally don't like the use of any model scoping for maintenance reasons but its the lesser of two abstraction evils.

I don't like this option because you're putting an if/else inside a scope, which doesn't actually make it more readable, so unless you add a scope for each case I'm not sure how this is in improvement.

Reply:
`so unless you add a scope for each case I'm not sure how this is in improvement.`

Look here how I've refactored it in the current version. I don't add a scope for each `where` case. The scope eliminates the case in the controller.

```ruby
 def show
    fetch_and_authorize_user

    target = @user
    target = @user.organization if @user&.organization && @user&.org_admin && params[:which] == "organization"
    @articles = target.articles.sorting(params[:sort]).decorate
    # Updates analytics in background if appropriate:
    ArticleAnalyticsFetcher.new.delay.update_analytics(current_user.id) if @articles
  end
```